Sources/AsyncHTTPClient/ConnectionsState.swift

```diff
@@ -151,37 +142,29 @@ extension HTTP1ConnectionProvider {
             }
         }
 
-        mutating func release(connection: Connection, closing: Bool) -> Action {
+        mutating func release(connection: ConnectionType, closing: Bool) -> Action<ConnectionType> {
             switch self.state {
             case .active:
                 assert(self.leasedConnections.contains(ConnectionKey(connection)))
 
                 if connection.isActiveEstimation, !closing { // If connection is alive, we can offer it to a next waiter
                     if let waiter = self.waiters.popFirst() {
+                        // There should be no case where we have both capacity and a waiter here.
+                        // Waiter can only exists if there was no capacity at aquire. If some connection
+                        // is released when we have waiter it can only indicate that we should lease (if EL are the same),
+                        // or replace (if they are different). But we cannot increase connection count here.
+                        assert(!self.hasCapacity)
+
                         let (eventLoop, required) = self.resolvePreference(waiter.preference)
 
                         // If returned connection is on same EL or we do not require special EL - lease it
-                        if connection.channel.eventLoop === eventLoop || !required {
+                        if connection.eventLoop === eventLoop || !required {
                             return .lease(connection, waiter)
                         }
 
-                        // If there is an opened connection on the same loop, lease it and park returned
-                        if let found = self.availableConnections.firstIndex(where: { $0.channel.eventLoop === eventLoop }) {
```

Comment:
the following two ifs were removed since they cannot be reached, we added and assert on line 156 instead